tmt/base.py

```diff
@@ -1796,7 +1797,14 @@ def _initialize_worktree(self) -> None:
 
         # Sync metadata root to the worktree
         self.debug(f"Sync the worktree to '{self.worktree}'.", level=2)
-        self.run(Command("rsync", "-ar", "--exclude", ".git", f"{tree_root}/", self.worktree))
+        excludes_tempfile = tempfile.NamedTemporaryFile()
+        # If we're in a git repository, honor .gitignore; xref
+        # https://stackoverflow.com/questions/13713101/rsync-exclude-according-to-gitignore-hgignore-svnignore-like-filter-c
+        if os.path.isdir(f"{tree_root}/.git"):
+            subprocess.check_call(["git", "ls-files", "--exclude-standard", "-oi", "--directory"], stdout=excludes_tempfile)
```

Comment:
* I'd recommend using either `Command` or `ShellScript`, that's how tmt runs external commands, adds logging and stuff. Given the output needs to end in a file, `ShellScript` seems better.
* it'd be nice to remove the temporary file once no longer used, I'm not 100 % sure but `with` should close & delete it.
* and we prefer `pathlib` instead of `os.path`.

Altogether, would something like this work?

```python
with tempfile.NamedTemporaryFile() as excludes_tempfile:
    if (tree_root / ".git").is_dir():
        self.run(ShellScript('git ls-files --exclude-standard -oi --directory > {excludes_tempfile.name}'))
    self.run(Command("rsync", "-ar", "--exclude", ".git", "--exclude-from", excludes_tempfile.name, f"{tree_root}/", self.worktree))
```

Reply:
OK, there were a few things that didn't "compile" there but I got to the point of this:

```
diff --git a/tmt/base.py b/tmt/base.py
index 7284a6ed..49f503f7 100644
--- a/tmt/base.py
+++ b/tmt/base.py
@@ -13,6 +13,7 @@ import sys
 import tempfile
 import time
 from collections.abc import Iterable, Iterator, Sequence
+from pathlib import Path
 from typing import (
     TYPE_CHECKING,
     Any,
@@ -1797,11 +1798,11 @@ class Plan(
 
         # Sync metadata root to the worktree
         self.debug(f"Sync the worktree to '{self.worktree}'.", level=2)
-        excludes_tempfile = tempfile.NamedTemporaryFile()
         # If we're in a git repository, honor .gitignore; xref
         # https://stackoverflow.com/questions/13713101/rsync-exclude-according-to-gitignore-hgignore-svnignore-like-filter-c
-        if os.path.isdir(f"{tree_root}/.git"):
-            subprocess.check_call(["git", "ls-files", "--exclude-standard", "-oi", "--directory"], stdout=excludes_tempfile)
+        with tempfile.NamedTemporaryFile() as excludes_tempfile:
+            if (Path(tree_root) / ".git").is_dir():
+                self.run(ShellScript(f"pwd; git ls-files --exclude-standard -oi --directory > {excludes_tempfile.name}").to_shell_command())
         # Note: rsync doesn't use reflinks right now, so in the future it'd be even better to
         # use e.g. `cp` but filtering out the above.
         self.run(Command("rsync", "-ar", "--exclude", ".git", "--exclude-from", excludes_tempfile.name, f"{tree_root}/", self.worktree))
```

Except for some reason I don't understand, the `ShellScript` variant tries to run with its working directory in `/var/tmp/tmt/run-015/plans/integration` and not the git repo...

Reply:
`self.run()` should accept `cwd` - no idea what the current CWD is, but we can tell the command to use a different one.

Importing `Path` should not be necessary, there's one imported already, from `tmt.utils` - it's our subclass of `pathlib.Path`, with few extra touches. Reading the code, `tree_root` should have been a `Path` instance already, you shouldn't need to coerce it in your patch, but it isn't: line 1790 should read `tree_root = Path(self.node.root) if self.node.root else None`, performing the conversion as soon as e path-like string enters tmt.

```python

Reply:
Would you mind taking over this PR?

Reply:
> Would you mind taking over this PR?

It's your idea, I don't want to step in & take the credit in git history or whatever changelog would happen to appear :) It would be nice to score new contributor from the family of tmt users, scratching their itch. If it looks like I'm always pointing out issues, sorry about that, it's not my intention to drive you away.

Reply:
I am not worried about credit, I would *much* rather be writing Rust code in bootc...
(If you do want to give credit though feel free to add [Co-authored-by](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors))

> If it looks like I'm always pointing out issues, sorry about that, it's not my intention to drive you away.


To be clear I think your suggestions are good, I just don't know the tmt codebase.

Reply:
@happz I belive this could be marked as resolved?